ocular/src/chain/client/transactions.rs

```diff
@@ -0,0 +1,170 @@
+use crate::error::TransactionError;
+use cosmrs::{
+    bank::MsgSend,
+    crypto::{secp256k1::SigningKey, PublicKey},
+    staking::{MsgDelegate, MsgUndelegate},
+    tendermint::chain::Id,
+    tx::{self, Fee, Msg, SignDoc, SignerInfo},
+    AccountId, Coin,
+};
+use tendermint_rpc::endpoint::broadcast::tx_commit::Response;
+
+use super::ChainClient;
+
+/// Metadata wrapper for transactions
+#[derive(Clone, Debug)]
+pub struct TransactionMetadata {
+    pub chain_id: Id,
+    pub account_number: u64,
+    pub sequence_number: u64,
+    pub gas_limit: u64,
+    pub timeout_height: u16,
+    pub memo: String,
+}
+
+impl ChainClient {
+    // TODO: Make this extensible to multisig and multicoin (or add new methods for that)
+    /// Helper method for signing and broadcasting messages.
+    async fn sign_and_send_msg(
+        &self,
+        sender_public_key: PublicKey,
+        sender_private_key: SigningKey,
+        amount: Coin,
+        tx_body: tx::Body,
+        tx_metadata: TransactionMetadata,
+    ) -> Result<Response, TransactionError> {
+        // Create signer info.
+        let signer_info =
+            SignerInfo::single_direct(Some(sender_public_key), tx_metadata.sequence_number);
+
+        // Compute auth info from signer info by associating a fee.
+        // TODO: Add ability to specify payers and granters for fees.
+        let auth_info =
+            signer_info.auth_info(Fee::from_amount_and_gas(amount, tx_metadata.gas_limit));
+
+        // Create doc to be signed
+        let sign_doc = match SignDoc::new(
+            &tx_body,
+            &auth_info,
+            &tx_metadata.chain_id,
+            tx_metadata.account_number,
+        ) {
+            Ok(doc) => doc,
+            Err(err) => return Err(TransactionError::TypeConversionError(err.to_string())),
+        };
+
+        // Create raw signed transaction.
+        let tx_signed = match sign_doc.sign(&sender_private_key) {
+            Ok(raw) => raw,
+            Err(err) => return Err(TransactionError::SigningError(err.to_string())),
+        };
+
+        // Broadcast transaction
+        match tx_signed.broadcast_commit(&self.rpc_client).await {
+            Ok(response) => Ok(response),
+            Err(err) => Err(TransactionError::BroadcastError(err.to_string())),
+        }
+    }
+
+    // TODO: Make this extensible to multisig and multicoin (or add new methods for that)
+    /// Signs and sends a simple transaction message.
+    pub async fn sign_and_send_msg_send(
```

Comment:
What do you think of naming this and the following two methods like this:

`send`
`delegate`
`undelegate`

Since the signing/sending is handled by `sign_and_send` it makes sense to me that these would have a higher-level name, and also the proto message that each deals with isn't something that really needs to be communicated in the method imo.

Reply:
That's fair, I felt like these method names might be a bit too verbose so i don't mind cutting it down to the core functionality; will update.

Reply:
Updated.